src/main/java/it/mulders/mcs/search/vulnerability/ComponentReportClient.java

```diff
@@ -0,0 +1,103 @@
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.util.Base64;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import com.github.packageurl.MalformedPackageURLException;
+import com.github.packageurl.PackageURL;
+import it.mulders.mcs.common.McsRuntimeException;
+import it.mulders.mcs.common.Result;
+import it.mulders.mcs.search.SearchResponse;
+
+public class ComponentReportClient {
+  private static final String MISSING_OSS_INDEX_CREDENTIALS_WARNING = """
+      You've requested to show reported security vulnerabilities, but you haven't configured a username and password
+      for the Sonatype OSS Index. See https://ossindex.sonatype.org for details on how this may impact your usage.""";
+
+  private final String hostname;
+  private final HttpClient client = HttpClient.newHttpClient();
+
+  public ComponentReportClient() {
+    this("https://ossindex.sonatype.org");
+  }
+
+  // Visible for testing
+  ComponentReportClient(final String hostname) {
+    this.hostname = hostname;
+  }
+
+  public Result<ComponentReportResponse> search(final SearchResponse.Response.Doc[] docs) {
+    return search(Stream.of(docs).map(this::toPackageUrls).toList());
+  }
+
+  public Result<ComponentReportResponse> search(final List<String> coordinates) {
+    var purls = coordinates.stream().map(purl -> "\"" + purl + "\"").collect(Collectors.joining(","));
+
+    var json = "{\"coordinates\" : [%s]}";
+
+    var builder = HttpRequest.newBuilder()
+    .version(HttpClient.Version.HTTP_1_1)
+    .POST(HttpRequest.BodyPublishers.ofString(json.formatted(purls)))
+    .header("Content-Type", "application/json");
+
+    var username = System.getProperty("ossindex.username", "");
+    var password = System.getProperty("ossindex.password", "");
+
+    if (!username.isEmpty() || !password.isEmpty()) {
+      builder.uri(URI.create(String.format("%s/api/v3/authorized/component-report", hostname)));
+      builder.header("Authorization", getBasicAuthenticationHeader(username, password));
+    } else {
+      builder.uri(URI.create(String.format("%s/api/v3/component-report", hostname)));
+      System.out.println(MISSING_OSS_INDEX_CREDENTIALS_WARNING);
+    }
+
+    try {
+      return client.send(builder.build(), new ComponentReportResponseBodyHandler()).body();
+    } catch (IOException | InterruptedException e) {
+      return new Result.Failure<>(e);
+    }
+  }
+
+  private String getBasicAuthenticationHeader(final String username, final String password) {
+    String valueToEncode = username + ":" + password;
+    return "Basic " + Base64.getEncoder().encodeToString(valueToEncode.getBytes());
+  }
+
+  public String toPackageUrls(final SearchResponse.Response.Doc doc) {
+    try {
+      return new PackageURL("maven", doc.g(), doc.a(), doc.v(), null, null).canonicalize();
```

Comment:
When the user performs a search that does not return one exact coordinate, `doc.v()` is empty. In that case, I think the best we could do is use `doc.latestVersion()`.

Reply:
> When the user performs a search that does not return one exact coordinate, `doc.v()` is empty. In that case, I think the best we could do is use `doc.latestVersion()`.

608650f3